File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/deeplink/DeepLinkActionTest.java
package com.emarsys.mobileengage.deeplink;

import android.app.Activity;
import android.content.Intent;

import com.emarsys.mobileengage.testUtil.TimeoutUtils;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

public class DeepLinkActionTest {

    static {
        mock(Intent.class);
        mock(Activity.class);
    }

    private DeepLinkInternal deepLinkInternal;
    private DeepLinkAction action;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void setUp() throws Exception {
        deepLinkInternal = mock(DeepLinkInternal.class);
        action = new DeepLinkAction(deepLinkInternal);

    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_mobileEngageInternalMustNotBeNull() {
        new DeepLinkAction(null);
    }

    @Test
    public void testExecute_callsMobileEngageInternal() throws Exception {
        Intent intent = mock(Intent.class);
        Activity activity = mock(Activity.class);
        when(activity.getIntent()).thenReturn(intent);

        action.execute(activity);

        verify(deepLinkInternal).trackDeepLinkOpen(activity, intent);
    }

    @Test
    public void testExecute_neverCallsMobileEngageInternal_whenIntentFromActivityIsNull() {
        Activity activity = mock(Activity.class);

        action.execute(activity);

        verifyZeroInteractions(deepLinkInternal);
    }

    @Test
    public void testExecute_neverCallsMobileEngageInternal_whenActivityIsNull() {
        action.execute(null);

        verifyZeroInteractions(deepLinkInternal);
    }

}

File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/deeplink/DeepLinkActionTest.kt
package com.emarsys.mobileengage.deeplink

import android.app.Activity
import android.content.Intent
import com.emarsys.mobileengage.di.setupMobileEngageComponent
import com.emarsys.mobileengage.di.tearDownMobileEngageComponent
import com.emarsys.mobileengage.fake.FakeMobileEngageDependencyContainer
import com.emarsys.mobileengage.util.waitForTask
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.After
import org.junit.Before
import org.junit.Test

class DeepLinkActionTest  {
    private lateinit var mockDeepLinkInternal: DeepLinkInternal
    private lateinit var action: DeepLinkAction

    @Before
    fun setUp() {

        setupMobileEngageComponent(FakeMobileEngageDependencyContainer())

        mockDeepLinkInternal = mockk(relaxed = true)
        action = DeepLinkAction(mockDeepLinkInternal)
    }

    @After
    fun tearDown() {
        tearDownMobileEngageComponent()
    }

    @Test
    fun testExecute_callsMobileEngageInternal() {
        val intent: Intent = mockk(relaxed = true)
        val mockActivity: Activity = mockk(relaxed = true)
        every { mockActivity.intent } returns intent

        action.execute(mockActivity)

        verify { mockDeepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null) }
    }

    @Test
    fun testExecute_neverCallsMobileEngageInternal_whenIntentFromActivityIsNull() {
        val mockActivity: Activity = mockk(relaxed = true)
        every { mockActivity.intent } returns null

        action.execute(mockActivity)
        waitForTask()

        verify(exactly = 0) {
            (mockDeepLinkInternal.trackDeepLinkOpen(
                eq(mockActivity),
                any(),
                any()
            ))
        }
    }

    @Test
    fun testExecute_neverCallsMobileEngageInternal_whenActivityIsNull() {
        action.execute(null)
        waitForTask()

        verify(exactly = 0) { (mockDeepLinkInternal.trackDeepLinkOpen(any(), any(), any())) }
    }
}